src/Features/CSharp/Portable/Completion/CompletionProviders/AwaitCompletionProvider.cs

```diff
@@ -0,0 +1,83 @@
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var declaration = root?.FindToken(completionListSpan.Start).GetAncestor(node => node.IsAsyncSupportingFunctionSyntax());
+            if (root is null || declaration is null)
+            {
+                return await base.GetChangeAsync(document, completionItem, completionListSpan, commitKey, disallowAddingImports, cancellationToken).ConfigureAwait(false);
+            }
+
+            // MethodDeclarationSyntax, LocalFunctionStatementSyntax, AnonymousMethodExpressionSyntax, ParenthesizedLambdaExpressionSyntax, SimpleLambdaExpressionSyntax.
+
+            var newDeclaration = AddAsyncModifier(declaration, SyntaxGenerator.GetGenerator(document));
+            var newDocument = document.WithSyntaxRoot(root.ReplaceNode(declaration, newDeclaration));
+            var changes = await newDocument.GetTextChangesAsync(document, cancellationToken).ConfigureAwait(false);
+            var newText = await newDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            return CompletionChange.Create(CodeAnalysis.Completion.Utilities.Collapse(newText, changes.ToImmutableArray()));
+        }
+
+        private static SyntaxNode AddAsyncModifier(SyntaxNode declaration, SyntaxGenerator generator)
+        {
+            var modifiers = generator.GetModifiers(declaration);
+            var newDeclaration = generator.WithModifiers(declaration, modifiers.WithAsync(true));
+            var asyncModifier = newDeclaration.GetModifiers().Single(m => m.IsKind(SyntaxKind.AsyncKeyword));
+            var newModifiers = newDeclaration.GetModifiers().Replace(asyncModifier, asyncModifier.WithAppendedTrailingTrivia(SyntaxFactory.ElasticSpace));
+            return newDeclaration.WithModifiers(newModifiers) ?? declaration;
+        }
+
+        public override async Task ProvideCompletionsAsync(CompletionContext context)
+        {
+            var document = context.Document;
+            var position = context.Position;
+            var cancellationToken = context.CancellationToken;
+
+            var semanticModel = await document.ReuseExistingSpeculativeModelAsync(position, cancellationToken).ConfigureAwait(false);
+
+            var workspace = document.Project.Solution.Workspace;
+            var syntaxContext = CSharpSyntaxContext.CreateContext(workspace, semanticModel, position, cancellationToken);
+
+            if (!syntaxContext.IsAwaitKeywordContext(position))
+            {
+                return;
+            }
+
+            var method = syntaxContext.TargetToken.GetAncestor(node => node.IsAsyncSupportingFunctionSyntax());
+            if (method is not null && !method.GetModifiers().Any(SyntaxKind.AsyncKeyword))
+            {
+                context.AddItem(CommonCompletionItem.Create("await", string.Empty, CompletionItemRules.Default, inlineDescription: "Make container async"));
```

Comment:
This might end up conflicting with the item provided by keyword recommender, and one would get deduplicated: since they have seem to have identical full display text (prefix + text + suffix), this will be decided by the order of providers being called.

See:
[http://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/Completion/CompletionServiceWithProviders.cs,484](http://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/Completion/CompletionServiceWithProviders.cs,484)

Reply:
Hiding the keyword might be the behavior we want, if so, we should figure out a way to make it explicit.
Otherwise, we need to show them both and give one higher priority . Before we get to the detail on how to implement this, let's figure out which approach we'd like to take.

---
In reply to: [517718403](https://github.com/dotnet/roslyn/pull/48352#discussion_r517718403) [](ancestors = 517718403)

Reply:
I don't think we have reached a conclusion on this yet. Current behavior relies on the implementation detail of completion service which is fragile. One way to resolve the interaction issue between two providers that provide "await" item is to fold them keyword provider into this one.